[MJ]미션3: 체스판 초기화

MJbae opened the pull request:
안녕하세요~ MJ입니다.
이번 리뷰도 잘 부탁드립니다!

**Console**
- 사용자 입력(start, end or else)에 따라 startGame(), endGame() 메소드를 실행합니다.
- startGame(): 체스게임을 초기화하고 현재 상태를 출력합니다.
- endGame(): 게임의 종료 메시지를 출력합니다.

**Board**
- initialize(): 내부 메소드 initPawns()를 호출하여 List pawns에 Pawn을 추가했습니다. 이어서 StringBuilder에 Pawn의 표현문자를 합치고 String[] cellsOfBoard에 StringBuilderd의 toString() 반환값을 삽입하였습니다.
- getRowOfBlackPawns(), getRowOfWhitePawns(): 초기화한 흰색폰과 검은색폰의 초기상태를 출력합니다.

**Pawn**
- 폰의 생성자에서 폰의 색깔을 결정하는 매개변수에 따라 표현문자(representation)도 결정합니다.

**BoardTest**
- initialize() 호출에 따라 흰색과 검은색 폰의 표현문자가 8개씩 합쳐져서 출력되는 지 확인합니다.

감사합니다~

Commit:
Test: assertAll 적용

하나의 테스트 메소드에 복수의 테스트가 포함되어 있는 경우,
실행순서를 기준으로 선순위의 테스트에 이상이 있으면 후순위의 테스트의 이상여부를 확인할 수 없음.
assertAll을 적용하면 실행순서에 관계없이 모든 테스트의 이상여부를 확인할 수 있음.

Commit:
Refactor: Pawn 객체 관리 콜렉션의 참조변수 타입을 List로 수정(ArrayList -> List)

모든 List collection class의 인터페이스인 List를 참조변수 타입으로 설정하여 다형성 추구
예를 들어, 참조변수 List pawns에 대해  ArrayList 외에도 LinkedList나 Stack 콜렉션을 사용할 수 있음

Commit:
Refactor: 클래스 내부에서 사용하는 메소드에 접근제어자 private 설정

테스트 내부에만 사용하기 때문에 코드의 보안성을 위해 외부의 접근을 제어함

Commit:
Refactor: piece의 색깔 타입을 제한하는 Enum의 이름 변경

Color라는 이름만으로 각 piece의 색에 대한 타입을 제한한다는 의미 전달할 수 있음
더불어 명료하기 때문에 코드 가독성을 높임

Commit:
Refactor Test: Enum의 클래스메소드 values() 활용하여 중복코드 제거

values()를 활용하여 모든 색깔(흰색과 검은색)에 대해 순회하여 테스트 가능
기존에 verifyColorOfPawn() 메소드를 각각의 색에 대해 두 번 명시했지만 수정 후 한 번만 명시

Commit:
Test: 폰 클래스의 표현문자 정상 출력 테스트 추가

폰 클래스 생성자의 두번째 매개변수로 표현문자 전달하고 해당 문자가 get method로 정상 출력되는지 확인

Commit:
Add: Pawn 객체 생성 시 표현문자값 전달

Pawn 객체에 상수로 표현문자를 정의하고 해당 문자를 생성자의 두번째 매개변수로 전달함

Commit:
Add: Pawn의 인스턴스 변수, representation에 대해 get method로 출력

Commit:
Refactor Test: 클래스 내부에서만 사용하는 메소드에 접근제어자 private 설정

Commit:
Test: initialize 메소드 호출에 따라 각 폰의 표현문자 정상출력 테스트 추가

Board class에서 initialize 호출하면 Pawn 객체의 인스턴스는 색깔별로 8개씩 생성되고 이를 합친 것이 정상적으로 출력되는지 확인

Commit:
Refactor Test: 불필요한 매개변수 전달 제거

특정 메소드에 고유하게 사용되는 값에 대해 매개변수로 전달하지 않고 지역변수로 사용하도록  수정

Commit:
Add: Board class에 initialize method 추가

해당 메소드 호출에 따라 List pawns에 흰색과 검은색의 폰이 각각 8개씩 추가됨
각각의 색의 폰의 표현문자에 대해 StringBuffer에 추가함

Commit:
Add: 체스게임의 사용자 인터페이스를 담당하는 Console class 추가

start, end라는 명령에 따라 게임을 시작하고 종료한다.
각각의 명령에 따라 간단하게 동작하는 임시 출력메소드를 사용했다

Commit:
Add: Board class에 체스판 초기화 결과를 출력하는 메소드 print() 추가

print()는 Board class의 initialize() method 호출에 따라 초기화된 board 상태를 출력한다.

Commit:
Add: Console class의 main method에 print method 사용

Commit:
Refactor: Console class 내 메소드를 기능에 따라 분할

Commit:
Refactor: Console class 내부에서만 사용하는 메소드에 대해 private 설정

Commit:
Refactor: Pawn 생성자에 매개변수로 color만 전달해도 해당 color에 맞는 representation 자동설정

Commit:
Refactor: initialize method의 불필요 코드 수정

while 반복문 안에서 두 번 명시되는 pawns.add를 한번만 명시함
변수 중에 상수로 전환해서 사용할 수 있는 것 변경

Commit:
Refactor: initialize method의 중복코드를 메소드 분할에 따라 제거
initialize method 안에 흰색폰과 검은색폰을 List pawns에 삽입하는 코드와 cellsOfBoard 배열에 흰색과 검은색의 representation의 모음을 삽입하는 코드가 중복됨
위의 중복되는 부분을 initPawns method로 분리하여 간소화함

Commit:
Refactor: 요구사항에 따라 ArrayList에 담긴 Pawn의 representation을 StringBuilder에 합친다

Commit:
Test: 흰색과 검은색 폰의 초기행의 결과를 출력하는 메소드의 이름 변경

LarryJung commented on src/main/java/mj/chess/Board.java:
```diff
@@ -1,11 +1,29 @@
+    private static final int SECOND_ROW = 1;
+    private static final int SEVENTH_ROW = 6;
+    private final List<Pawn> pawns = new ArrayList<>();
+    private final StringBuilder rowOfBlackPawns = new StringBuilder();
+    private final StringBuilder rowOfWhitePawns = new StringBuilder();
```
만약.. 체스게임을 진행하면서 기물(현재는 pawn)의 색깔이 바뀐다던지? 위치가 바뀌었을때? rowOfBlackPawns과 같은 출력물에 대한 것도 다시 업데이트를 해주어야겠죠? 

그런 관점에서 Board가 가지고 있어야 할 아주 코어한 데이터가 어떤 이유에서 변동이 있을때마다 관련된 필드들(rowOfBlackPawns, rowOfWhitePawns, cellsOfBoard) 들을 업데이트 하는 것은 손이 많이 갈 것 같기도 한데요,

혹시 아래 print(), getRowOfBlackPawns(), getRowOfWhitePawns() 를 호출할 때 이미 만들어놓은 값을 활용하여 심플하게 리턴하지 않고 메서드 호출 시점에 가진 기물을 가지고 프린트하는 로직을 돌리는 것은 생각해 보셨나요?

OOP는 왠지 관련된 데이터를 클래스라는 한 묶음으로 두고 최대한 일을 많이 시키는게 중요합니다.ㅎㅎ

static final 필드 제외하고 멤버 변수를 pawns 만 둘 수 밖에 없다면 어떻게 짤 수 있을지 고민해 보세요!

MJbae replied:
리뷰에 따라 제 코드를 붙잡고 고민해보니, List pawns만으로 구현할 수 있는 로직에 불필요한 변수를 사용하는 부분이 많았습니다. 
List pawns를 기반으로 stream을 활용하여 Board getRowOfPawns()의 내부 로직을 개선하였습니다.

```java
public String getRowOfPawns(Color color) {
    return pawns.stream()
            .filter(pawn -> pawn.getColor() == color)
            .map(Pawn::getRepresentation)
            .map(Object::toString)
            .collect(joining());
}
```

LarryJung commented on src/main/java/mj/chess/Board.java:
```diff
@@ -21,4 +39,31 @@ public int size() {
         return pawns.get(index);
     }
+
+    public void initialize() {
+        initPawns(Color.WHITE, Pawn.WHITE_REPRESENTATION, SECOND_ROW, rowOfWhitePawns);
+        initPawns(Color.BLACK, Pawn.BLACK_REPRESENTATION, SEVENTH_ROW, rowOfBlackPawns);
+    }
+
+    private void initPawns(Color color, char representation, final int NTH_ROW, StringBuilder rowOfPawns) {
+        for (int i = 0; i < MAX_SIZE; i++) {
+            pawns.add(new Pawn(color));
+            rowOfPawns.append(representation);
+        }
+        cellsOfBoard[NTH_ROW] = (color == Color.BLACK) ? getRowOfBlackPawns() : getRowOfWhitePawns();
+    }
```
리팩토링을 해보자면.. 

```suggestion
    public void initialize() {
        initPawns(Color.WHITE);
        initPawns(Color.BLACK);
        initCellsOfBoard();
    }

    private void initPawns(Color color, int nthRow) {
        for (int i = 0; i < MAX_SIZE; i++) {
            pawns.add(new Pawn(color));
        }
    }
    
    private void initCellsOfBoard() {
      cellsOfBoard[WHITE_PAWN_ROW] = pawns.stream().filter(p -> p.isColor(Color.WHITE)).map(p -> p.getRepre()).joinString() 
      cellsOfBoard[BLACK_PAWN_ROW] = pawns.stream().filter(p -> p.isColor(Color.BLACK)).map(p -> p.getRepre()).joinString()
    }
```

코드로 직접 리뷰를 드리기보다 생각할 것을 던져드리려고 했는데.. 쓰다보니 코드를 적어버렸네요 ㅎㅎ
제가 하고 싶은 말은 initPawns메서드가 일단 두가지 일을 하고 있는 것으로 보였구요,

1. pawns을 넣는 부분
2.cells 초기화 하는 부분

기능 분리를 좀 해봤고.. (여기서 힌트는 메서드 인자가 3,4개 이상 넘어가고 있다면 어딘지 쓰이는 곳이 많다는 사인일 수도 있습니다.) 

그리고 initPawns 메서드에서 color가 지정된 Pawn의 getRepresentation() 메서드가 있음에도 initPawns에 char 를 인자로 넘겨서 pawn 객체를 충분히 활용하고 있지 못한 점이 아쉽네요 ㅠ 

제가 예시로 코드를 드렸는데 한번 살펴봐 주시고, 어떤 장,단점이 있을지 의견을 주시면 좋을것 같습니다.!

(코드는 좀 의사(psuedo)코드라 정확하지 않습니다. 정답도 아니구요!)

LarryJung replied:
추가 코멘트 드리면 파라미터 컨벤션을 지켜주세요!

MJbae replied:
본 리뷰에서 Larry가 강조하는 건 코드의 재사용성으로 이해했습니다. 코드의 재사용성을 높이는 것의 장점을 세 가지로 정리했습니다. 
1. 성능향상, JVM의 Execution Engine 중 JIT Compiler는 한 번 읽어드린 코드에 대해서는 line by line이 아니라 한 번에 실행합니다. 따라서 재사용되는 코드가 많을 수록 프로그램 전체의 성능은 향상됩니다.
2. 수정작업 시 효율성 증가, 코드의 재사용률이 높으면 수정사항이 발생해도 개발자가 일일이 변경할 코드의 양은 줄어들 것입니다.
3. 에러 발생 감소, 사람이 개발하는 이상 인간적인 실수가 발생할 수 있습니다. 사람이 일일이 건드는 변수의 양을 줄이고 반복적으로 사용하는 코드가 많아진다면 인간적인 실수도 함께 줄 것입니다.

단점은 아무래도 재사용을 하기 위해 다른 클래스의 코드를 가져와야 하기 때문에 코드의 간결성이 떨어진다는 것으로 볼 수 있습니다. 하지만 이것도 어떻게 재사용하는지에 따라서 오히려 간결해질 수 있습니다.

LarryJung commented on src/main/java/mj/chess/Board.java:
```diff
@@ -21,4 +39,31 @@ public int size() {
+        }
+    }
+
+    public String getRowOfBlackPawns() {
+        return rowOfBlackPawns.toString();
+    }
+
+    public String getRowOfWhitePawns() {
+        return rowOfWhitePawns.toString();
+    }
```
위에도 언급드렸듯이 요때 pawns를 가지고 로직을 돌린다면..? rowOfBlackPawns 멤버 변수는 없어도 되지 않을까요?

LarryJung replied:
아 그리고 내부에서 사용하고 있으니 private으로 바꿔주세요~

MJbae replied:
1. 해당 메소드를 public으로 설정한 이유는 테스트 코드에서 사용하기 때문입니다. 
2. pawns에 스트림을 활용하니까 rowOfBlackPawns가 필요 없어졌습니다! ㅎㅎ

MJbae replied:
혹시 Larry의 의도는 테스트에 사용할 메소드의 경우 package-private으로 설정하라는 것으로 이해해도 될까요?

LarryJung replied:
아 넵 테스트에서 사용한다면 package-private으로 둬도 괜찮습니다.^^

LarryJung commented on src/main/java/mj/chess/Board.java:
```diff
@@ -21,4 +39,31 @@ public int size() {
+        cellsOfBoard[NTH_ROW] = (color == Color.BLACK) ? getRowOfBlackPawns() : getRowOfWhitePawns();
+    }
+
+    public void print() {
+        for (String rowOfBoard : cellsOfBoard) {
+            System.out.println(rowOfBoard);
+        }
+    }
```
print() 가 프린트를 하고 있지만, 혹시 println 은 콘솔쪽으로 위임하고, Board는 문자열만 리턴하게 하면 어떤 장점이 있을까요?

MJbae replied:
코드관리에 이점이 있습니다. 출력도 하나의 기능이므로 출력을 전담하는 객체를 두는 것이 추후 전체적인 코드 관리면에 이롭습니다. 출력부분을 Board에서 Console로 이동했습니다.

LarryJung commented on src/main/java/mj/chess/pieces/Pawn.java:
```diff
@@ -1,18 +1,33 @@
 
     public Pawn() {
-        this.color = ColorOfPiece.WHITE;
+        this.color = Color.WHITE;
+        this.representation = WHITE_REPRESENTATION;
+    }
+
+    public Pawn(Color color) {
+        this.color = color;
+        this.representation = (color == Color.BLACK) ? BLACK_REPRESENTATION : WHITE_REPRESENTATION;
```
현재 요구사항에서는 문제가 없지만! Board와 비슷하게 representation을 멤버 변수로 두지 않고, getRepresentation() 그때그때 계산해서 값을 반환한다면 color가 바뀌어도 다른 로직을 수정할 필요가 없다는게 장점일거라 생각이 드네요~

어떻게 하는게 더 코드관리에 용이할지 생각해봐주세요 ㅎㅎ

MJbae replied:
코드의 재사용성을 높이기 위해 getRepresentation()을 사용하는 것이 적절하겠네요!

LarryJung commented on src/main/java/mj/chess/pieces/Pawn.java:
```diff
@@ -1,18 +1,33 @@
+        return representation;
+    }
+
+}
```
파일 끝 개행 추가해주세요!

MJbae replied:
개행문자를 추가하는 이유는 POSIX 명세, 일종의 표준을 지켜서 다른 프로그램과의 호환 시 에러의 발생 가능성을 제거하기 위해서군요!

LarryJung reviewed (changes requested):
수고하셨습니다~! 요구사항 대로 잘 구현하셨구요, 이런저런 코멘트를 좀 드렸으니 확인 부탁드립니다. :)

Commit:
Refactor: initPawns()에 전달되는 매개변수 중 char representation 제거

char representation은 Pawn class에 정의된 getPresentation()으로 대체할 수 있음.
관리할 변수를 최대한 줄이면 코드 관리가 쉬워지고 예상치 못한 에러발생을 방지할 수 있음.

Commit:
Refacotr: 매개변수 컨벤션에 따라 매개변수 이름 변경

Commit:
Refactor: initPawns() 내부에 Pawns외 cellsOfBoard에 대한 초기화 부분 분할

메소드 하나에는 하나의 기능을 포함한다는 원칙에 따라 두 가지 기능(pawns, cellsOfBoard 각각에 대한 초기화) 중 cellsOfBoard
의 초기화에 대한 메소드를 새로 정의

Commit:
Refactor: StringBuilder rowOfPawns의 기능을 stream으로 대체

List pawns의 표현문자를 문자열로 다루기 이해 StringBuilder rowOfPawns를 사용했음
하지만, 추가적인 변수를 선언하지 않고도 List pawns에 스트림을 적용하면 위의 기능을 구현할 수 있음
스트림을 활용하는 방법이 코드의 재사용성을 높이고 불필요한 변수를 제거하기 때문에 코드의 관리면에서 더욱 효과적임

Commit:
Refactor Test: getRowOfPawns에 스트림 적용에 따라 불필요한 코드중복 제거

기존 테스트 코드는 흰색폰과 검은색폰 각각의 표현문자의 출력기능에 대해 각각의 메소드를 선언했음.
하지만, 스트림 활용에 따라 메소드로 전달할 매개변수가 줄어들었고 하나의 메소드로 통합할 수 있었음

Commit:
Refactor: Enum values() 활용하여 코드중복 제거

Enum의 타입제한 활용하여 Board의 모든 메소드의 매개변수로 Enum만 전달함
Enum values() 활용하여 흰색과 검은색에 따라 두 번 호출했던 부분을 한 번 호출하는 것으로 수정

Commit:
Refactor: 파일 끝 개행 추구

파일 끝에 개행문자를 추가하는 이유는 POSIX 명세를 따르기 위한 것
일종의 표준으로 지키지 않으면 다른 프로그램과의 호완 시에 에러가 발생할 수 있음

MJbae commented:
@LarryJung Larry의 리뷰에 감사합니다. 머리로 이해만 하고 있던 부분을 실제 코드로 구현해보면서 체득할 수 있었습니다.

LarryJung commented on src/main/java/mj/chess/Board.java:
```diff
@@ -66,11 +68,11 @@ public void print() {
+    public String getRowOfPawns(Color color) {
+        return pawns.stream()
+                .filter(pawn -> pawn.getColor() == color)
+                .map(Pawn::getRepresentation)
+                .map(Object::toString)
```
map 을 두번 연속으로 하게 되면 객체 생성 비용이 들기 때문에 map 은 하나도 모아주는것이 좋습니다^^

MJbae replied:
그렇군요! 스트림을 더 공부해봐야겠습니다!

LarryJung commented on src/main/java/mj/chess/Board.java:
```diff
@@ -58,8 +58,12 @@ private void initPawns(Color color) {
-    private void initCellsOfBoard(Color color, final int nthRow) {
-        cellsOfBoard[nthRow] = getRowOfPawns(color);
+    private void initCellsOfBoard(Color color) {
+        if (color == Color.WHITE) {
+            cellsOfBoard[SECOND_ROW] = getRowOfPawns(color);
+        } else {
+            cellsOfBoard[SEVENTH_ROW] = getRowOfPawns(color);
+        }
```
나중에 남몰래 Color가 추가된다면 이쪽에 원치않는 결과가 나올 수도 있는데, switch 를 통해 white, black 을 확실하게 정해주는것은 어떨까요?

MJbae replied:
else로 넘기지 말고 확실하게 조건을 걸고 처리하라는 말씀이시군요! 생각해보니 Color를 건든다면 의도치 않은 결과로 이어질 수 있겠네요! 좋은 지적 감사합니다 :)

LarryJung commented:
코드 수정하시느라 고생하셨습니다.^^ 제 기준에는 많이 간결해지고 가독성도 좋아진 느낌인데 어떻게 생각하실 지 모르겠네요 ㅎㅎ 
마이너한 리뷰 추가로 드렸으니 참고하시고 다음단계 넘어가셔도 될듯 합니다~

MJbae commented:
> 코드 수정하시느라 고생하셨습니다.^^ 제 기준에는 많이 간결해지고 가독성도 좋아진 느낌인데 어떻게 생각하실 지 모르겠네요 ㅎㅎ
> 마이너한 리뷰 추가로 드렸으니 참고하시고 다음단계 넘어가셔도 될듯 합니다~

Larry 도움으로 이전에 고려하지 않았던 부분을 고민하면서 코드를 작성하게 되었습니다. 특히 코드의 재사용과 변경사항에 대처할 수 있는 여지를 남겨야한다는 것이 이번 리뷰에서 가장 크게 배운 부분입니다. 감사합니다~
